### ISSUE TITLE
AR::Relation cache_key_with_version remains same when updated_at has changed

### ISSUE THREAD

**Issue by okuramasafumi** [rails/rails#41784]
### Steps to reproduce

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", path: "./"
  # gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveRecord::Base.collection_cache_versioning = true

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.timestamps
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
    t.timestamps
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_cache_key
    post = Post.create!
    post.comments << Comment.create!

    cache_key = post.comments.cache_key_with_version
    comment = post.comments.first
    comment.touch(:updated_at, time: Time.new(2021, 5, 1))
    new_cache_key = post.comments.cache_key_with_version
    refute_equal cache_key, new_cache_key
  end
end
```

### Expected behavior

`cache_key_with_version` is different since max value of `updated_at` in the collection has changed.

### Actual behavior

`cache_key_with_version` remains the same.

### System configuration
**Rails version**: 6.1.3.1

**Ruby version**: 2.7.2

### related issues/prs

https://github.com/rails/rails/pull/34378

**Comment by ghiculescu** [comment]
It looks like https://github.com/rails/rails/pull/41789 fixes this and github just didn't close it automatically. Will close the issue, but please reopen if that's incorrect. cc @kamipo

**Comment by lazyatom** [comment]
I've just run the included reproduction script against Rails 6.1, 7.0 and git HEAD, and the error is reproduced in all cases.

It may be impossible (or impractical) to fix the relation cache key if `touch` is called on a member, but I am confident that #41789 doesn't actually fix this; it would only address it if the collection is updated via a method like `update_all`. #41789 doesn't address any other ways of updating this, and also only addresses `cache_key`, and not `cache_version` or `cache_key_with_version`.
